flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -1882,9 +1882,11 @@ int getNumQueuedRequests() {
     public void reportStats(long id, ExecutionAttemptID attemptId, CheckpointMetrics metrics)
             throws CheckpointException {
         if (statsTracker != null) {
-            attemptMappingProvider
-                    .getVertex(attemptId)
-                    .ifPresent(ev -> statsTracker.reportIncompleteStats(id, ev, metrics));
+            synchronized (lock) {
```

Comment:
In fact, I think we should check nullability of `statsTracker` under the lock as well since that variable is not volatile. But as I can see we already have a place where we don't do that `trackPendingCheckpointStats` so I suppose we have some happens-before protection here. but if it is not true we should fix both places.

Reply:
If I understand the code correctly the `statsTracker` is set when creating the `CheckpointCoordinator`. Tbh, I believe it should be passed in the ctor and honestly it might never be null. It might've been nullable in the past. 

WDYT about passing the `statsTracker` in the ctor, removing the setter and making it non null? See `org.apache.flink.runtime.executiongraph.DefaultExecutionGraph#enableCheckpointing`. It is the only place we're constructing the `CheckpointCoordinator` and the `statsTracker` can not be null there (line 408).

Reply:
It seems we can do it